src/dotty/tools/dotc/core/Definitions.scala

```diff
@@ -99,8 +99,12 @@ class Definitions {
   lazy val RootPackage: TermSymbol = ctx.newSymbol(
     NoSymbol, nme.ROOTPKG, PackageCreationFlags, TypeRef(NoPrefix, RootClass))
 
-  lazy val EmptyPackageClass = ctx.newCompletePackageSymbol(RootClass, nme.EMPTY_PACKAGE).moduleClass.asClass
-  lazy val EmptyPackageVal = EmptyPackageClass.sourceModule.entered
+  lazy val EmptyPackageVal = ctx.newCompletePackageSymbol(RootClass, nme.EMPTY_PACKAGE).entered
+  lazy val EmptyPackageClass = EmptyPackageVal.moduleClass.asClass
+
+  /** A package in which we can place all methods that are interpreted specially by the compiler */
```

Comment:
Yes, sure, Scala.js can call JS. But JS code is opaque to the optimizer, so it cannot inline these methods. And there are some things the JS code *cannot* do, for example get/set/length of an array, because a Scala Array is opaque to JS.

Intercepted methods (intrinsics) are only intercepted when the optimizer is enabled. It is not acceptable to *require* it. Even identityHashCode has a valid implementation in Scala.js code, which is used when the optimizer is not there to intercept it.

So there's no way to do this but for the back-end to capture these method calls as primitives. Primitives are different from intrinsics in that they *have* to be intercepted by the back-end and transformed into special IR nodes. Intrinsics are normal method calls, with a valid implementation, but which can optionally be intercepted by the optimizer to use a more efficient implementation.